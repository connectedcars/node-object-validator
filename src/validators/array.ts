import { ValidatorBase } from '../common'
import { NotArrayError, RequiredError, ValidationErrorContext, WrongLengthError } from '../errors'
import { ValidatorTypes } from '../types'

export function validateArray(
  value: unknown,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error[] | null {
  if (!Array.isArray(value)) {
    return [new NotArrayError(`Must be an array (received "${value}")`, context)]
  }
  if ((minLength !== 0 && value.length < minLength) || value.length > maxLength) {
    return [
      new WrongLengthError(
        `Must contain between ${minLength} and ${maxLength} entries (found ${value.length})`,
        context
      )
    ]
  }
  return null
}

export class RequiredArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'RequiredArray' = 'RequiredArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return [new RequiredError(`Is required`, context)]
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export class OptionalArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'OptionalArray' = 'OptionalArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return null
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required?: false
): OptionalArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required: true
): RequiredArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  required = false
): OptionalArray<T> | RequiredArray<T> {
  return required
    ? new RequiredArray<T>(schema, minLength, maxLength)
    : new OptionalArray<T>(schema, minLength, maxLength)
}
