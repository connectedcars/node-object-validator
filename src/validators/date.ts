import { CodeGenResult, ValidatorBase, ValidatorBaseOptions, ValidatorExportOptions, ValidatorOptions } from '../common'
import { NotDateFail, RequiredFail, ValidationFailure } from '../errors'

export function isDate(value: unknown, context?: string): value is Date {
  const errors = validateDate(value, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateDate(value: unknown, context?: string): ValidationFailure[] {
  if (!(value instanceof Date)) {
    return [new NotDateFail(`Must be a Date object`, value, context)]
  }
  return []
}

export abstract class DateValidator<O = never> extends ValidatorBase<Date | O> {
  public constructor(options?: ValidatorBaseOptions) {
    super(options)
    if (options?.optimize !== false) {
      this.optimize()
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
     ...this.nullCheckWrap([
      `  if (!(${localValueRef} instanceof Date)) {`,
      `    errors.push(new NotDateFail(\`Must be a Date object\`, ${localValueRef}${contextStr}))`,
      `  }`,
      ], localValueRef, contextStr),
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        NotDateFail: NotDateFail,
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string): ValidationFailure[] {
    return validateDate(value, context)
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = `Date`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {

        const typeStr = 'std::time::Instant'
        const isOption = !this.required || this.nullable
        return isOption ? `Option<${typeStr}>` : `${typeStr}`
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredDate extends DateValidator {
  public constructor(options?: ValidatorOptions) {
    super({ ...options })
  }
}

export class OptionalDate extends DateValidator<undefined> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false })
  }
}

export class NullableDate extends DateValidator<null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, nullable: true })
  }
}

export class OptionalNullableDate extends DateValidator<undefined | null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false, nullable: true })
  }
}
