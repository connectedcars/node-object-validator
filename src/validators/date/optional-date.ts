import { ValidationErrorContext } from '../../errors'
import { validateDate } from './validate-date'

export class OptionalDate {
  private type: 'OptionalDate' = 'OptionalDate'

  public validate(value: Date, context?: ValidationErrorContext<string>): Error | null {
    if (value == null) {
      return null
    }
    return validateDate(value, context)
  }
}
