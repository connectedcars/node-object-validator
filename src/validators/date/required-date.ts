import { RequiredError, ValidationErrorContext } from '../../errors'
import { validateDate } from './validate-date'

export class RequiredDate {
  private type: 'RequiredDate' = 'RequiredDate'

  public validate(value: Date, context?: ValidationErrorContext<string>): Error | null {
    if (value == null) {
      return new RequiredError(`Is required`, context)
    }
    return validateDate(value, context)
  }
}
