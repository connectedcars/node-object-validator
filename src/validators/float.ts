import { ValidatorBase } from '../common'
import { NotFloatError, OutOfRangeError, RequiredError, ValidationErrorContext } from '../errors'

export function validateFloat(
  value: unknown,
  min = Number.MIN_SAFE_INTEGER,
  max = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error[] {
  if (typeof value !== 'number' || isNaN(value) || !isFinite(value)) {
    return [new NotFloatError(`Must be a float (received "${value}")`, context)]
  }
  if (value < min || value > max) {
    return [new OutOfRangeError(`Must be between ${min} and ${max} (received "${value}")`, context)]
  }
  return []
}

export class RequiredFloat extends ValidatorBase<number> {
  private type: 'RequiredFloat' = 'RequiredFloat'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    super()
    this.min = min
    this.max = max
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] {
    if (value == null) {
      return [new RequiredError(`Is required`, context)]
    }
    return validateFloat(value, this.min, this.max, context)
  }
}

export class OptionalFloat extends ValidatorBase<number | undefined | null> {
  private type: 'OptionalFloat' = 'OptionalFloat'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    super()
    this.min = min
    this.max = max
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] {
    if (value == null) {
      return []
    }
    return validateFloat(value, this.min, this.max, context)
  }
}

export function Float(min: number, max: number, required?: false): OptionalFloat
export function Float(min: number, max: number, required: true): RequiredFloat
export function Float(min = 0, max = Number.MAX_SAFE_INTEGER, required = false): OptionalFloat | RequiredFloat {
  return required ? new RequiredFloat(min, max) : new OptionalFloat(min, max)
}
