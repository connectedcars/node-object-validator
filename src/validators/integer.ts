import { NotIntegerError, OutOfRangeError, RequiredError, ValidationErrorContext } from '../errors'
import { Validator } from './common'

export function validateInteger(
  value: number,
  min = 0,
  max = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error | null {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    return new NotIntegerError(`Must be an integer (received "${value}")`, context)
  }
  if (value < min || value > max) {
    return new OutOfRangeError(`Must be between ${min} and ${max} (received "${value}")`, context)
  }
  return null
}

export class RequiredInteger implements Validator {
  private type: 'RequiredInteger' = 'RequiredInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return new RequiredError(`Is required`, context)
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export class OptionalInteger implements Validator {
  private type: 'OptionalInteger' = 'OptionalInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return null
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export function Integer(required?: false): OptionalInteger
export function Integer(required: true): RequiredInteger
export function Integer(required = false): OptionalInteger | RequiredInteger {
  return required ? new RequiredInteger() : new OptionalInteger()
}
