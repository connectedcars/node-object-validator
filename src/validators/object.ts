import {
  CodeGenResult,
  ValidateOptions,
  ValidatorBase,
  ValidatorBaseOptions,
  ValidatorExportOptions,
  ValidatorOptions
} from '../common'
import { NotObjectFail, RequiredFail, ValidationFailure } from '../errors'
import { toSnakeCase } from '../util'

export function isObject<T extends ObjectSchema>(
  schema: T,
  value: unknown,
  context?: string
): value is ObjectWrap<UndefinedToOptional<{ [K in keyof T]: T[K]['tsType'] }>> {
  const errors = validateObject(schema, value, context, { earlyFail: true })
  if (errors.length === 0) {
    return true
  }
  return false
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isPlainObject(value: unknown): value is Record<string | number | symbol, any> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

export function validateObject(
  schema: Record<string, ValidatorBase>,
  value: unknown,
  context?: string,
  options?: ValidateOptions
): ValidationFailure[] {
  const errors: ValidationFailure[] = []
  if (!isPlainObject(value)) {
    errors.push(new NotObjectFail(`Must be an object`, value, context))
    return errors
  }
  for (const key of Object.keys(schema)) {
    const validator = schema[key]
    const keyName = context ? `${context}['${key}']` : key
    errors.push(...validator.validate(value[key], keyName, { optimized: false, earlyFail: false, ...options }))
    if (options?.earlyFail && errors.length > 0) {
      return errors
    }
  }
  return errors
}

export type ObjectSchema = Record<string, ValidatorBase>

// https://github.com/Microsoft/TypeScript/issues/26705
export type IsUndefined<T, K> = undefined extends T ? K : never
export type IsNotUndefined<T, K> = undefined extends T ? never : K
export type UndefinedKeys<T> = { [K in keyof T]-?: IsUndefined<T[K], K> }[keyof T]
export type NotUndefinedKeys<T> = { [K in keyof T]-?: IsNotUndefined<T[K], K> }[keyof T]
export type IncludeUndefinedTypes<T extends Record<string, unknown>> = { [K in UndefinedKeys<T>]: T[K] }
export type ExcludeUndefinedTypes<T extends Record<string, unknown>> = { [K in NotUndefinedKeys<T>]: T[K] }
export type UndefinedToOptional<T extends Record<string, unknown>> = ExcludeUndefinedTypes<T> &
  Partial<IncludeUndefinedTypes<T>>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ObjectWrap<T> = T extends Record<string, any> ? { [K in keyof T]: T[K] } : never

export abstract class ObjectValidator<T extends ObjectSchema = never, O = never> extends ValidatorBase<
  ObjectWrap<UndefinedToOptional<{ [K in keyof T]: T[K]['tsType'] }>> | O
> {
  public schema: ObjectSchema
  private typeStringGenerated: boolean

  public constructor(schema: T, options?: ValidatorBaseOptions) {
    super(options)
    this.typeStringGenerated = false
    this.schema = schema
    if (options?.optimize !== false) {
      this.optimize(schema)
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const objValueRef = `objValue${id()}`
    const schemaRef = `scheme${id()}`
    let imports: { [key: string]: unknown } = {
      NotObjectFail: NotObjectFail,
      RequiredFail: RequiredFail
    }
    const declarations = [`const ${schemaRef} = ${validatorRef}.schema`]

    const propValidationCode: string[] = []
    for (const key of Object.keys(this.schema)) {
      const validator = this.schema[key]
      const propName = context ? `${context}['${key}']` : key
      const [propImports, propDeclarations, propCode] = validator.codeGen(
        `${objValueRef}['${key}']`,
        `${schemaRef}['${key}']`,
        id,
        propName,
        earlyFail
      )
      imports = { ...imports, ...propImports }
      declarations.push(...propDeclarations)
      propValidationCode.push(...propCode)
    }

    // prettier-ignore
    const code = [
      `const ${objValueRef} = ${valueRef}`,
      ...this.nullCheckWrap([
      `  if (typeof ${objValueRef} === 'object' ${!this.nullable ? `&& ${objValueRef} !== null ` : ''}&& !Array.isArray(${objValueRef})) {`,
      ...propValidationCode.map(l => `    ${l}`),
      `  } else {`,
      `    errors.push(new NotObjectFail(\`Must be an object\`, ${objValueRef}${contextStr}))`,
      `  }`,
      ], objValueRef, contextStr),
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]

    return [imports, declarations, code]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString(options)
    }
  }

  protected validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    return validateObject(this.schema, value, context, { earlyFail: this.earlyFail, ...options })
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        const lines = Object.keys(this.schema).map(k =>
          `'${k}': ${this.schema[k].toString(options)}`.replace(/(^|\n)/g, '$1  ')
        )
        let typeStr = `{\n${lines.join('\n')}\n}`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        if (options?.rustTypeName === undefined) {
          throw new Error(`'rustTypeName' option is not set`)
        }
        if (!this.typeStringGenerated) {
          this.typeStringGenerated = true

          const lines = Object.keys(this.schema).map(k => `  ${toSnakeCase(k)}: ${this.schema[k].toString(options)},`)
          return `struct ${options?.rustTypeName} {\n${lines.join('\n')}\n}`
        }
        const isOption = !this.required || this.nullable
        return isOption ? `Option<${options.rustTypeName}>` : `${options.rustTypeName}`
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(options?: ValidatorExportOptions): string {
    const lines = Object.keys(this.schema).map(k =>
      `'${k}': ${this.schema[k].toString(options)}`.replace(/(^|\n)/g, '$1  ')
    )
    const schemaStr = `{\n${lines.join(',\n')}\n}`
    const optionsStr = this.optionsString !== '' ? `, ${this.optionsString}` : ''
    return `new ${this.constructor.name}(${schemaStr}${optionsStr})`
  }
}

export class RequiredObject<T extends ObjectSchema> extends ObjectValidator<T> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options })
  }
}

export class OptionalObject<T extends ObjectSchema> extends ObjectValidator<T, undefined> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false })
  }
}

export class NullableObject<T extends ObjectSchema> extends ObjectValidator<T, null> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, nullable: true })
  }
}

export class OptionalNullableObject<T extends ObjectSchema> extends ObjectValidator<T, undefined | null> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false, nullable: true })
  }
}
