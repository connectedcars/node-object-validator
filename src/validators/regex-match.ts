import { CodeGenResult, isValidType, ValidatorBase, ValidatorOptions } from '../common'
import {
  DoesNotMatchRegexFail,
  NotStringFail,
  RequiredFail,
  ValidationErrorContext,
  ValidationFailure,
  WrongLengthFail
} from '../errors'
import { validateString } from './string'

export function isRegexMatch<T extends string>(
  value: unknown,
  regex: RegExp,
  context?: ValidationErrorContext
): value is T {
  const errors = validateRegexMatch(value, regex, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateRegexMatch(
  value: unknown,
  regex: RegExp,
  context?: ValidationErrorContext
): ValidationFailure[] {
  const stringError = validateString(value, 0, Number.MAX_SAFE_INTEGER, context)
  if (!isValidType<string>(value, stringError)) {
    return stringError
  }
  if (!regex.test(value)) {
    return [new DoesNotMatchRegexFail(`Did not match '${regex}' (received "${value}")`, context)]
  }
  return []
}

export class RegexMatchValidator<O = never> extends ValidatorBase<string | O> {
  private regex: RegExp

  public constructor(regex: RegExp, options?: ValidatorOptions) {
    super(options)
    this.regex = regex
    if (options?.optimize) {
      this.optimize()
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: ValidationErrorContext,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, { key: \`${context.key}\` }` : ', context'
    const localRegexRef = `regex${id()}`
    const localValueRef = `value${id()}`
    // prettier-ignore
    const declarations: string[] = [
      `const ${localRegexRef} = ${this.regex}`
    ]
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      `if (${localValueRef} != null) {`,
      `  if (typeof ${localValueRef} === 'string') {`,
      `    if (!${localRegexRef}.test(${localValueRef})) {`,
      `      errors.push(new DoesNotMatchRegexFail(\`Did not match '${this.regex}' (received "\${${localValueRef}}")\`${contextStr}))`,
      `    }`,
      `  } else {`,
      `    errors.push(new NotStringFail(\`Must be a string (received "\${${localValueRef}}")\`${contextStr}))`,
      `  }`,
      ...(this.required ? [
      `} else {`,
      `  errors.push(new RequiredError(\`Is required\`${contextStr}))`] : []),
      '}',
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),

    ]
    return [
      {
        DoesNotMatchRegexFail: DoesNotMatchRegexFail,
        WrongLengthFail: WrongLengthFail,
        NotStringFail: NotStringFail,
        RequiredError: RequiredFail
      },
      declarations,
      code
    ]
  }

  protected validateValue(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    return validateRegexMatch(value, this.regex, context)
  }
}

export class RequiredRegexMatch extends RegexMatchValidator<string> {
  public constructor(regex: RegExp, options?: ValidatorOptions) {
    super(regex, { ...options, required: true })
  }
}

export class OptionalRegexMatch extends RegexMatchValidator<undefined | null> {
  public constructor(regex: RegExp, options?: ValidatorOptions) {
    super(regex, { ...options, required: false })
  }
}
