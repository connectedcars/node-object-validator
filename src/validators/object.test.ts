/* eslint-disable @typescript-eslint/no-unused-vars */
import { NotIntegerFail, RequiredFail } from '../errors'
import { AssertEqual } from '../types'
import { OptionalArray, RequiredArray } from './array'
import { OptionalDate } from './date'
import { RequiredFloat } from './float'
import { OptionalInteger, RequiredInteger } from './integer'
import { ObjectValidator, OptionalObject, RequiredObject, TypedObject, validateObject } from './object'
import { RegexMatch, RequiredRegexMatch } from './regex-match'

describe.each([false, true])('Object (optimize: %s)', optimize => {
  describe('validateObject', () => {
    it('should validate simple object', () => {
      const errors = validateObject(
        new RequiredObject({
          int: new RequiredInteger(1, 2)
        }),
        { int: 1 }
      )
      expect(errors).toEqual([])
    })
  })

  describe('ObjectValidator', () => {
    it('should validation and give expected result', () => {
      const objectValidator = new ObjectValidator(
        {
          int: new RequiredInteger(1, 2),
          optionalInt: new OptionalInteger(1, 2),
          requiredObject: new RequiredObject({
            int: new RequiredInteger(1, 2),
            optionalInt: new OptionalInteger(1, 2)
          }),
          optionalArray: new OptionalArray(new RequiredInteger(1, 2)),
          optionalArrayArray: new OptionalArray(new RequiredArray(new RequiredInteger(1, 2))),
          optionalDate: new OptionalDate(),
          regexMatch: new RequiredRegexMatch(/^.*$/)
        },
        { optimize }
      )

      const unknownValue: unknown = {
        int: 1,
        optionalInt: 1,
        requiredObject: {
          int: 1
        },
        optionalArray: [1],
        optionalArrayArray: [[1]],
        optionalDate: new Date(),
        regexMatch: 'hello'
      }
      const str = objectValidator.validate.toString()
      if (optimize) {
        expect(str).toMatch(/generatedFunction = true/)
      } else {
        expect(str).not.toMatch(/generatedFunction = true/)
      }
      const errors = objectValidator.validate(unknownValue)
      expect(errors).toEqual([])
    })

    it('should cast type guard correctly for isValid', () => {
      const objectValidator = new ObjectValidator(
        {
          int: new RequiredInteger(1, 2),
          float: new RequiredFloat(1, 2),
          optionalInt: new OptionalInteger(1, 2),
          requiredObject: new RequiredObject({
            int: new RequiredInteger(1, 2),
            optionalInt: new OptionalInteger(1, 2)
          }),
          optionalArray: new OptionalArray(new RequiredInteger(1, 2)),
          optionalArrayArray: new OptionalArray(new RequiredArray(new RequiredInteger(1, 2)))
        },
        { optimize }
      )

      const unknownValue: unknown = {
        int: 1,
        float: 1.5,
        optionalInt: 1,
        requiredObject: {
          int: 1
        },
        optionalArray: [1],
        optionalArrayArray: [[1]]
      }
      if (objectValidator.isValid(unknownValue)) {
        const itShouldCastIntToNumber: AssertEqual<typeof unknownValue.int, number> = true
        const itShouldCastFloatToNumber: AssertEqual<typeof unknownValue.float, number> = true
      } else {
        expect('did not validate but should').toBe('')
      }
    })

    it('should cast to known type', () => {
      const objectValidator = new ObjectValidator(
        {
          int: new RequiredInteger(1, 2)
        },
        { optimize }
      )

      const unknownValue: unknown = {
        int: 1
      }
      expect(() => {
        const knownValue = objectValidator.cast<{ int: number }>(unknownValue)
        const itShouldCastIntToNumber: AssertEqual<typeof knownValue.int, number> = true
      }).not.toThrow()
    })

    it('should fail to cast', () => {
      const objectValidator = new ObjectValidator(
        {
          int: new RequiredInteger(1, 2)
        },
        { optimize }
      )

      const unknownValue: unknown = {
        int: '1'
      }
      expect(() => {
        const knownValue = objectValidator.cast(unknownValue)
        const itShouldCastIntToNumber: AssertEqual<typeof knownValue.int, number> = true
      }).toThrow()
    })

    it('should reject non object values', () => {
      const validator = new ObjectValidator({}, { optimize })
      expect(validator.validate([]).map(e => e.toString())).toStrictEqual([
        'NotObjectFail: Must be an object (received "")'
      ])
      expect(validator.validate(1).map(e => e.toString())).toStrictEqual([
        'NotObjectFail: Must be an object (received "1")'
      ])
      expect(validator.validate(true).map(e => e.toString())).toStrictEqual([
        'NotObjectFail: Must be an object (received "true")'
      ])
      expect(validator.validate('').map(e => e.toString())).toStrictEqual([
        'NotObjectFail: Must be an object (received "")'
      ])
    })
  })

  describe('RequiredObject', () => {
    const objectValidator = new RequiredObject(
      {
        int: new RequiredInteger(1, 2),
        optionalInt: new OptionalInteger(1, 2),
        requiredObject: new RequiredObject({
          int: new RequiredInteger(1, 2),
          optionalInt: new OptionalInteger(1, 2)
        }),
        optionalArray: new OptionalArray(new RequiredInteger(1, 2)),
        optionalArrayArray: new OptionalArray(new RequiredArray(new RequiredInteger(1, 2)))
      },
      { optimize }
    )

    it('should export the correct type and compile', () => {
      const itShouldAllowOptionalParameters: typeof objectValidator.type = {
        int: 0,
        optionalInt: 1,
        requiredObject: {
          int: 0
        },
        optionalArray: [0],
        optionalArrayArray: [[0]]
      }

      const itShouldCastIntToNumber: AssertEqual<typeof objectValidator.type.int, number> = true
      const itShouldCastOptionalIntToNumberOrUndefined: AssertEqual<
        typeof objectValidator.type.optionalInt,
        number | undefined
      > = true
      const itShouldCastOptionalArrayToNumberArrayOrUndefined: AssertEqual<
        typeof objectValidator.type.optionalArray,
        number[] | undefined
      > = true
    })

    it('accepts empty value', function () {
      const validator = new RequiredObject({}, { optimize })
      expect(validator.validate(null)).toStrictEqual([new RequiredFail('Is required')])
      expect(validator.validate(undefined as unknown)).toStrictEqual([new RequiredFail('Is required')])
    })

    it('validates correct value', () => {
      const unknownValue: unknown = {
        int: 1,
        optionalInt: 1,
        requiredObject: {
          int: 1
        },
        optionalArray: [1],
        optionalArrayArray: [[1]]
      }
      expect(objectValidator.validate(unknownValue)).toEqual([])
    })

    it('rejects wrong types', () => {
      const unknownValue: unknown = {
        int: 1,
        optionalInt: 1,
        requiredObject: {
          int: 1,
          optionalInt: '1'
        },
        optionalArray: ['1'],
        optionalArrayArray: [1]
      }
      expect(objectValidator.validate(unknownValue)).toEqual([
        new NotIntegerFail(`Must be an integer (received "1")`, {
          key: "requiredObject['optionalInt']"
        }),
        new NotIntegerFail(`Must be an integer (received "1")`, {
          key: 'optionalArray[0]'
        }),
        new NotIntegerFail(`Must be an array (received "1")`, {
          key: 'optionalArrayArray[0]'
        })
      ])
    })

    it('should cast type guard correctly for isType', () => {
      const unknownValue: unknown = {
        int: 1,
        optionalInt: 1,
        requiredObject: {
          int: 1
        },
        optionalArray: [1],
        optionalArrayArray: [[1]]
      }
      const errors = objectValidator.validate(unknownValue)
      if (objectValidator.isType(unknownValue, errors)) {
        const itShouldCastIntToNumber: AssertEqual<typeof unknownValue.int, number> = true
      } else {
        expect('did not validate but should').toBe('')
      }
    })

    it('should cast type guard correctly for cast', () => {
      const unknownValue: unknown = {
        int: 1,
        optionalInt: 1,
        requiredObject: {
          int: 1
        },
        optionalArray: [1],
        optionalArrayArray: [[1]]
      }
      const knownValue = objectValidator.cast(unknownValue)
      const itShouldCastIntToNumber: AssertEqual<typeof knownValue.int, number> = true
    })
  })

  describe('OptionalObject', () => {
    it('accepts empty value', function () {
      const validator = new OptionalObject({}, { optimize })
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
      const knownValue = validator.cast(null)
      const itShouldCastIntToNumber: AssertEqual<typeof knownValue, {} | null | undefined> = true
    })
  })

  describe('TypedObject', () => {
    it('accepts empty value', () => {
      const validator = TypedObject({}, false)
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
    })

    it('rejects empty value', () => {
      const validator = TypedObject({})
      expect(validator.validate(null).map(e => e.toString())).toStrictEqual(['RequiredFail: Is required'])
      expect(validator.validate(undefined).map(e => e.toString())).toStrictEqual(['RequiredFail: Is required'])
    })
  })
})
