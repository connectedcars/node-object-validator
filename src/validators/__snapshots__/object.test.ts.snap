// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object RequiredInteger should return a function body 1`] = `
Array [
  Object {
    "NotIntegerFail": [Function],
    "NotObjectFail": [Function],
    "OutOfRangeFail": [Function],
    "RequiredFail": [Function],
  },
  Array [
    "const scheme2 = validator1.schema",
  ],
  Array [
    "const objValue1 = value1",
    "if (objValue1 !== undefined) {",
    "  if (typeof objValue1 === 'object' && objValue1 !== null && !Array.isArray(objValue1)) {",
    "    const value3 = objValue1['int']",
    "    if (value3 !== undefined) {",
    "      if (typeof value3 === 'number' && Number.isInteger(value3)) {",
    "        if (value3 < 1 || value3 > 2) {",
    "          errors.push(new OutOfRangeFail(\`Must be between 1 and 2\`, value3, \`int\`))",
    "        }",
    "      } else {",
    "        errors.push(new NotIntegerFail(\`Must be an integer\`, value3, \`int\`))",
    "      }",
    "    } else {",
    "      errors.push(new RequiredFail(\`Is required\`, value3, \`int\`))",
    "    }",
    "  } else {",
    "    errors.push(new NotObjectFail(\`Must be an object\`, objValue1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, objValue1, context))",
    "}",
  ],
]
`;
